ref: corrections to `exp run/apply/branch` + (#2417)
iterative/dvc.org: content/docs/command-reference/exp/apply.md

## Description

Restores a given `experiment` (name or hash are accepted, see `dvc exp run` for
details) into the workspace. This means changing all files (code, data,
<abbr>parameters</abbr>, <abbr>metrics</abbr>, etc.) needed to reflect the
experiment conditions and results in the <abbr>workspace</abbr>.

This command is typically used after finding a target experiment with
`dvc exp show` or `dvc exp diff`, and before committing it to Git (making it
[persistent](/doc/user-guide/experiment-management#persistent-experiments)).

Note that this command may fail if the current Git commit (`HEAD`) has changed

> significantly from the time the `experiment` was run.

Reviewer: It fails outright if your current `HEAD` commit does not match the `HEAD` commit from when you ran the experiment.

It may also fail if there are git conflicts between changes you have in your workspace and changes in the experiment you want to apply. We try to automatically resolve the conflicts by preferring the changes in the applied experiment, but sometimes git may still fail to auto-resolve them.

> significantly from the time the `experiment` was run.

Author: Updated in 8762f974. PTAL

> We try to automatically resolve the conflicts by preferring the changes in the applied experiment

Meaning `git merge --theirs`, right @pmrowla ? I think it's not considered a conflict until after Git can't merge it so I went with the phrase "if they conflict", for now.

> Where  they have moved their HEAD ... and then try to exp apply some experiment

Does/should the CLI print a HINT to try `exp branch` instead?

> significantly from the time the `experiment` was run.

Reviewer: No, it will give an error along the lines of "cannot apply `exp-...` onto commit `<user's current HEAD>`. Experiment is derived from `<original experiment HEAD>`"

> significantly from the time the `experiment` was run.

Author: > exp branch is for a different use case than exp apply

OK @pmrowla. Would you mind checking the bullet list in the latest version to confirm that it covers the use case(s)? Thanks

## Description

> Restores a given `experiment` (name or hash are accepted, see `dvc exp run` for

Reviewer 2: Could we make the statement in parentheses a new sentence?

Restores a given `experiment` into the workspace. Experiment can be referenced by a hash or name.

Author: Doesn't seem like a sentence pertaining to that paragraph to me. Would have to be another p just for the argument (and maybe some of the options), which I'm trying to avoid.

Author: I agree it would be easier to read though...

details) into the workspace. This means changing all files (code, data,
<abbr>parameters</abbr>, <abbr>metrics</abbr>, etc.) needed to reflect the
experiment conditions and results in the workspace.

This command is typically used after finding a target `experiment` with
`dvc exp show` or `dvc exp diff`, and before committing it to Git (making it
[persistent](/doc/user-guide/experiment-management#persistent-experiments)).

> Note that any [checkpoints](/doc/command-reference/exp/run#checkpoints) found in

Reviewer 3: ```suggestion
Note that any intermediate [checkpoints](/doc/command-reference/exp/run#checkpoints) found in
```
Wondering if `intermediate` better conveys the message that the whole history is "squashed" into a single merge rather than disregarded. `apply` will still reflect the experiment state at the time of the final checkpoint. Feel free to ignore if you think it's unhelpful.

Author: Yeah it's more correct, but not really a concept we have explained much. I don't know if it's intuitive enough. I went for "the history of checkpoints" for now.

> the `experiment` are **not** preserved when applying and committing it.

Reviewer 3: Alternate suggestion (same idea as previous comment adding `intermediate`):
```suggestion
the `experiment` are **not** preserved in separate commits when applying.
```
Again, not a blocker, decide for yourself.
